### PR TITLE
Fix #2564: Do not create ill-formed <init> accessors

### LINKED ISSUE #2564: Inlined private constructor crashes in GenBCode
```scala
object Foo {
  inline def bar = new Bar
  class Bar private[Foo]()
}
```

with `Xprint` crashes with
```scala
result of tests/run/Foo.scala after TreeTransform:{moveStatic, expandPrivate, selectStatic, collectEntryPoints, collectSuperCalls, dropInlined, labelDef}:

package <empty> {
  @scala.annotation.internal.SourceFile("tests/run/Foo.scala") final module 
    class
   Foo$ extends Object { 
    def <init>(): Unit = 
      {
        super()
        ()
      }
    @Foo.bar$_inlineAccessor_$1(new Foo.Foo$Bar)() inline def bar(): Foo.Foo$Bar
       = 
    Foo.bar$_inlineAccessor_$1(new Foo.Foo$Bar)
    def bar$_inlineAccessor_$1(x$0: Foo.Foo$Bar): Foo.Foo$Bar = x$0()
  }
  class Foo$Bar extends Object { 
    def <init>(): Unit = 
      {
        super()
        ()
      }
  }
  final lazy module val Foo: Foo$ = new Foo$()
}

Unexpected New(Foo.Bar/TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),Foo$)),Bar)) reached GenBCode.
  Call was genLoad(New(TypeTree[TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),Foo$)),Bar)]),LFoo$Bar;)
java.lang.RuntimeException: Unexpected New(Foo.Bar/TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),Foo$)),Bar)) reached GenBCode.
  Call was genLoad(New(TypeTree[TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),Foo$)),Bar)]),LFoo$Bar;)
	at dotty.tools.backend.jvm.DottyBackendInterface.abort(DottyBackendInterface.scala:381)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:312)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder$$anonfun$genLoadArguments$1.apply(BCodeBodyBuilder.scala:1004)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder$$anonfun$genLoadArguments$1.apply(BCodeBodyBuilder.scala:1004)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoadArguments(BCodeBodyBuilder.scala:1004)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genNormalMethodCall$1(BCodeBodyBuilder.scala:725)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genApply(BCodeBodyBuilder.scala:754)
	at scala.tools.nsc.backend.jvm.BCodeBodyBuilder$PlainBodyBuilder.genLoad(BCodeBodyBuilder.scala:337)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.emitNormalMethodBody$1(BCodeSkelBuilder.scala:606)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genDefDef(BCodeSkelBuilder.scala:640)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:509)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder$$anonfun$gen$1.apply(BCodeSkelBuilder.scala:511)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder$$anonfun$gen$1.apply(BCodeSkelBuilder.scala:511)
	at scala.collection.immutable.List.foreach(List.scala:392)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.gen(BCodeSkelBuilder.scala:511)
	at scala.tools.nsc.backend.jvm.BCodeSkelBuilder$PlainSkelBuilder.genPlainClass(BCodeSkelBuilder.scala:118)
	at dotty.tools.backend.jvm.GenBCodePipeline$Worker1.visit(GenBCode.scala:199)
	at dotty.tools.backend.jvm.GenBCodePipeline$Worker1.run(GenBCode.scala:152)
	at dotty.tools.backend.jvm.GenBCodePipeline.buildAndSendToDisk(GenBCode.scala:363)
	at dotty.tools.backend.jvm.GenBCodePipeline.run(GenBCode.scala:329)
	at dotty.tools.backend.jvm.GenBCode.run(GenBCode.scala:54)
	at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:283)
	at dotty.tools.dotc.core.Phases$Phase$$anonfun$runOn$1.apply(Phases.scala:281)
	at scala.collection.immutable.List.map(List.scala:284)
	at dotty.tools.dotc.core.Phases$Phase$class.runOn(Phases.scala:281)
	at dotty.tools.backend.jvm.GenBCode.runOn(GenBCode.scala:38)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:82)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1$$anonfun$apply$mcV$sp$1.apply(Run.scala:79)
	at scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)
	at scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply$mcV$sp(Run.scala:79)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:67)
	at dotty.tools.dotc.Run$$anonfun$compileUnits$1.apply(Run.scala:67)
	at dotty.tools.dotc.util.Stats$.monitorHeartBeat(Stats.scala:76)
	at dotty.tools.dotc.Run.compileUnits(Run.scala:67)
	at dotty.tools.dotc.Run.compileSources(Run.scala:64)
	at dotty.tools.dotc.Run.compile(Run.scala:48)
	at dotty.tools.dotc.Driver.doCompile(Driver.scala:26)
	at dotty.tools.dotc.Driver.process(Driver.scala:124)
	at dotty.tools.dotc.Driver.process(Driver.scala:93)
	at dotty.tools.dotc.Driver.process(Driver.scala:105)
	at dotty.tools.dotc.Driver.main(Driver.scala:132)
	at dotty.tools.dotc.Main.main(Main.scala)

Error while emitting Foo.scala
Unexpected New(Foo.Bar/TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),Foo$)),Bar)) reached GenBCode.
  Call was genLoad(New(TypeTree[TypeRef(ThisType(TypeRef(ThisType(TypeRef(NoPrefix,<empty>)),Foo$)),Bar)]),LFoo$Bar;)
```

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
This just a first small step to ensure that the compiler does not crash on inlined private constructors.

File: compiler/src/dotty/tools/dotc/typer/Inliner.scala [hunk 1]
```diff
@@ -144,9 +144,15 @@ object Inliner {
           case _: Apply | _: TypeApply | _: RefTree if needsAccessor(tree.symbol) =>
             if (tree.isTerm) {
               val (methPart, targs, argss) = decomposeCall(tree)
-              addAccessor(tree, methPart, targs, argss,
-                  accessedType = methPart.tpe.widen,
-                  rhs = (qual, tps, argss) => qual.appliedToTypes(tps).appliedToArgss(argss))
+              val mSym = methPart.symbol
+              if (mSym.isConstructor && (mSym.is(Private) || mSym.privateWithin.exists)) {
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala]: Do we need to check `privateWithin`, is it subsumed by the `private` check?

Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala]: I thought so too but when trying it out I found that it was not the case. Should it be? Did I stumble on another bug?

Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala]: The `privateWithin` check is fine, https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/typer/Inliner.scala#L49.

Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/typer/Inliner.scala]: Actually I should use `mSym.isConstructor && needsAccessor`

**Pull request merged**